src/js/context/ContextProvider.ts

```diff
@@ -0,0 +1,188 @@
+
+        this.setIsEdited = this.setIsEdited.bind(this);
+        this.getData = this.getData.bind(this);
+        this.setData = this.setData.bind(this);
+        this.transformData = this.transformData.bind(this);
+        this.yieldData = this.yieldData.bind(this);
+    }
+
+    static getConstructorConfig(config: ContextProviderConfig): ContextProviderInstance {
+        return {
+            constructor: this.prototype.constructor as typeof ContextProvider,
+            config,
+        };
+    }
+
+    static createConfigFromContext(config: ContextProviderConfig) {
+        const data = lodash.get(config.context, config.name);
+        const isEdited = lodash.get(config.context, this.getIsEditedKeyByName(config.name));
+        const extraData = lodash.get(config.context, this.getExtraDataKeyByName(config.name));
+        return Object.assign(
+            config,
+            data
+                ? {
+                      data,
+                      extraData,
+                      isEdited,
+                  }
+                : {},
+        );
+    }
+
+    setIsEdited(isEdited: boolean) {
+        this.isEdited = isEdited;
+    }
+
+    getData() {
+        return this.isEdited ? this.data : this.defaultData;
+    }
+
+    setData(data: object) {
+        this.data = deepClone(data);
+    }
+
+    // override in children
+    // eslint-disable-next-line class-methods-use-this
+    get defaultData() {
+        throw new Error("Not implemented.");
+    }
+
+    // override in children
+    // eslint-disable-next-line class-methods-use-this
+    transformData(data: object) {
+        return data;
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    yieldData(...transformDataArgs: any) {
+        const extraDataObject = this.extraData ? { [this.extraDataKey]: this.extraData } : {};
+        return {
+            ...extraDataObject,
+            // @ts-ignore
+            [this.name]: this.transformData(this.getData(), ...transformDataArgs),
+            [this.isEditedKey]: this.isEdited,
+        };
+    }
+
+    // override when this.data needs additional processing before making it available to rendering context
+    // used to calculate explicit points path, for example
+    yieldDataForRendering() {
+        return this.yieldData();
+    }
+
+    get extraDataKey() {
+        return `${this.name}ExtraData`;
+    }
+
+    static getExtraDataKeyByName(name: string) {
+        return `${name}ExtraData`;
+    }
+
+    get isEditedKey() {
+        return `is${lodash.capitalize(this.name)}Edited`;
+    }
+
+    static getIsEditedKeyByName(name: string) {
+        return `is${lodash.capitalize(name)}Edited`;
+    }
+
+    get isUnitContextProvider() {
+        return this.entityName === "unit";
+    }
+
+    get isSubworkflowContextProvider() {
+        return this.entityName === "subworkflow";
+    }
```

Comment:
Just in general - this class has a lot of functionality through its functions and we have very little if any unit tests for it - we should have more tests here to be able to track any changes before they get to the webapp

Reply:
Added unit tests for this class in the next PR